nodkz/mongodb-memory-server
# fix/improveMongoInstance

- rename interface "MongodOps" to "MongodOpts" to keep naming consistency
- remove static value "childProcessList" because it was never used
- move "dynamic" function "debug" into private class function
- shorten "constructor" to use less lines
- outsource "MongodOpts.instance" into own interface "MongoInstanceOpts"
- change root values of "MongodOpts" to be required
- remove unused return value inside "run"'s "launch" Promise
- add tsdoc to "kill"
- add tsdoc to "errorHandler"
- extend tsdoc for "closeHandler"
- remove value "opts"
- add value "instanceOpts"
- add value "binaryOpts"
- add value "spawnOpts"
- prepareCommandArgs: remove splitting "this.instanceOpts"
- prepareCommandArgs: convert values to booleans (to ensure no falsy values)
- improve tsdoc for "waitPrimaryReady"
- change "MongoInstance" to extend "EventEmitter"
- add various Events
- track event names with an enum
- handle not being PRIMARY anymore
- "_waitForPrimary" use new events instead of function
- remove function "waitPrimaryReady"
- remove value "waitForPrimaryResolveFns"
- replace instanceReady with Event "instanceReady"
- replace instanceFailed with Event "instanceError"
- remove case for "shutting down with code"
- update case for "address already in use"
- add listener to "run" on "instanceClosed"
- test the error value on reject
- change test name to be more readable and accurate
- feat(db_util): add function "assertion"
- add error if "post" is null/undefined in "prepareCommandArgs" (because that is the purpose if this package)
- add error if "dbPath" is null/undefined in "prepareCommandArgs" (because that is the purpose if this package)
- move "ip" inside an if
- enforce "stdio" to be always "pipe"
- try "SIGKILL" if process didnt exit within 10 seconds
- reject promise if process dosnt exit after sending "SIGKILL"
- rename parameter "process" to "childProcess" to not overwrite global "process"
- move function "kill_internal" from "MongoInstance" to "db_util"
- rename function "kill_internal" to "killProcess"
- log the array created in the function
- overwrite "emit", "on", "once" to only allow "MongoInstanceEvents" instead of "string | symbol"
- add "@fires" to all functions that emit an event
- test that "kill" dosnt call "killProcess" if no childProcesses are running
- test that "_launchMongod" throws if "childProcess.pid" is undefined
- test that "errorHandler" emits "instanceRawError" and "instanceError"
- test that "stderrHandler" emits "instanceSTDERR"
- test that "stdoutHandler" emits correct event on "waiting for connections"
- move all "event testing" tests into "test events"
- add "beforeAll" creating the spy & instance & events map
- add "beforeEach" resetting spy-calls & events map
- test that "stdoutHandler" emits correct event on "address already in use"
- test that "stdoutHandler" emits correct event on "CURL_OPENSSL_3"
- test that "stdoutHandler" emits correct event on "CURL_OPENSSL_4"
- test that "stdoutHandler" emits correct event on "transition primary complete"
- test that "stdoutHandler" emits correct event on "state change"
- add "TODO" for events in "stdoutHandler" in "MongoInstance" that arnt covered yet
- remove redundant "debug.enabled" check
- rename event "instanceState" to "instanceReplState"
- remove "| null" from "childProcess"
- remove "| null" from "killerProcess"
- change "childProcess" to be optional
- change "killerProcess" to be optional
- reset "childProcess" reference to "undefined" in "stop"
- reset "killerProcess" reference to "undefined" in "stop"
- test(replset-multi): remove variable "opts"
- add tsdoc to the variables in "Mongoinstance"

## Related Issues

- closes #365
- most likely fixes #166 

---

because this pr includes breaking changes, i thought to bring this together with other file-improvements
